File: server/sso.js
// ĐĂNG NHẬP MICROSOFT (OpenID Connect — Authorization Code + PKCE)
//
// Vì sao tự viết thay vì thêm thư viện: luồng này chỉ cần 2 request HTTP + 1 lần verify JWT.
// Node 20 đã có sẵn `fetch` và `crypto` (đọc được JWK trực tiếp), `jsonwebtoken` thì repo đã dùng.
// Thêm @azure/msal-node kéo theo hàng chục phụ thuộc vào một app đang giữ nguyên tắc "vanilla, ít dep".
//
// 4 lớp chống giả mạo, THIẾU MỘT LÀ THỦNG:
//   state         — chống CSRF: kẻ khác không ép được trình duyệt bạn "đăng nhập" bằng mã của hắn
//   nonce         — chống phát lại: id_token cũ bắt được không dùng lại được
//   PKCE (S256)   — chống cướp mã: bắt được `code` cũng vô dụng nếu không có code_verifier
//   verify chữ ký — id_token phải do CHÍNH tenant đó ký (JWKS), đúng iss/aud/exp/tid
//
// state + nonce + code_verifier KHÔNG lưu ở server (app stateless) mà gói trong một JWT ngắn hạn
// đặt ở cookie httpOnly riêng — hết luồng là xoá.
const crypto = require('crypto');
const jwt = require('jsonwebtoken');
const { getSettings } = require('./db');
const { JWT_SECRET } = require('./auth');

const STATE_COOKIE = 'ktx_sso';
const STATE_TTL_SEC = 600;                 // 10 phút: đủ cho người dùng gõ mật khẩu + 2FA, không hơn
const b64url = buf => buf.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');

// Cấu hình: ENV được ƯU TIÊN hơn CSDL. Môi trường thật nên giữ bí mật ở ENV; ô trong màn Cài đặt
// là để chạy thử/khi công ty chưa cấp được ENV. Không đủ tham số -> coi như TẮT (nút Microsoft tự ẩn).
async function ssoConfig() {
  let s = {};
  try { s = await getSettings(); } catch (e) { s = {}; }
  const pick = (env, key) => (process.env[env] || s[key] || '').trim();
  const tenantId = pick('AZURE_TENANT_ID', 'sso_tenant_id');
  const clientId = pick('AZURE_CLIENT_ID', 'sso_client_id');
  const clientSecret = pick('AZURE_CLIENT_SECRET', 'sso_client_secret');
  const domains = pick('SSO_ALLOWED_DOMAINS', 'sso_allowed_domains')
    .split(',').map(d => d.trim().toLowerCase().replace(/^@/, '')).filter(Boolean);
  // Bật khi: được bật tường minh (ENV hoặc Cài đặt) VÀ đủ 3 tham số bắt buộc.
  const on = (process.env.SSO_ENABLED || s.sso_enabled || 'false') === 'true';
  return { enabled: !!(on && tenantId && clientId && clientSecret), tenantId, clientId, clientSecret, allowedDomains: domains };
}

const issuerOf = tenantId => `https://login.microsoftonline.com/${tenantId}/v2.0`;
const authorizeEndpoint = tenantId => `https://login.microsoftonline.com/${tenantId}/oauth2/v2.0/authorize`;
const tokenEndpoint = tenantId => `https://login.microsoftonline.com/${tenantId}/oauth2/v2.0/token`;
const jwksUri = tenantId => `https://login.microsoftonline.com/${tenantId}/discovery/v2.0/keys`;

// Địa chỉ quay về sau khi Microsoft xác thực xong. PHẢI trùng tuyệt đối với Redirect URI
// đã khai trong Azure. Cho phép ép bằng ENV vì sau proxy (Render) đôi khi đoán sai scheme/host.
function redirectUri(req) {
  if (process.env.SSO_REDIRECT_URI) return process.env.SSO_REDIRECT_URI.trim();
  const proto = (req.headers['x-forwarded-proto'] || req.protocol || 'http').split(',')[0].trim();
  return `${proto}://${req.get('host')}/api/auth/sso/callback`;
}

// ---- JWKS: khoá công khai của Microsoft để kiểm chữ ký id_token (cache 1 giờ) ----
let _jwksCache = { tenantId: null, at: 0, keys: [] };
async function getSigningKey(tenantId, kid) {
  const fresh = _jwksCache.tenantId === tenantId && (Date.now() - _jwksCache.at) < 3600e3;
  if (!fresh) {
    const r = await fetch(jwksUri(tenantId));
    if (!r.ok) throw new Error('Không lấy được khoá công khai của Microsoft (JWKS)');
    const j = await r.json();
    _jwksCache = { tenantId, at: Date.now(), keys: j.keys || [] };
  }
  let jwk = _jwksCache.keys.find(k => k.kid === kid);
  if (!jwk) {                                  // Microsoft xoay khoá -> nạp lại đúng một lần
    const r = await fetch(jwksUri(tenantId));
    if (r.ok) { const j = await r.json(); _jwksCache = { tenantId, at: Date.now(), keys: j.keys || [] }; }
    jwk = _jwksCache.keys.find(k => k.kid === kid);
  }
  if (!jwk) throw new Error('Không tìm thấy khoá ký khớp id_token');
  return crypto.createPublicKey({ key: jwk, format: 'jwk' });
}

// ---- Bước 1: dựng URL đẩy người dùng sang Microsoft ----
async function buildAuthRequest(req) {
  const cfg = await ssoConfig();
  if (!cfg.enabled) throw Object.assign(new Error('Đăng nhập Microsoft chưa được cấu hình'), { status: 503 });
  const state = b64url(crypto.randomBytes(24));
  const nonce = b64url(crypto.randomBytes(24));
  const codeVerifier = b64url(crypto.randomBytes(48));
  const codeChallenge = b64url(crypto.createHash('sha256').update(codeVerifier).digest());
  const uri = redirectUri(req);
  const params = new URLSearchParams({
    client_id: cfg.clientId,
    response_type: 'code',
    redirect_uri: uri,
    response_mode: 'query',
    scope: 'openid profile email',
    state, nonce,
    code_challenge: codeChallenge,
    code_challenge_method: 'S256',
  });
  // Gói bí mật của luồng vào JWT ngắn hạn -> cookie httpOnly. Không lưu gì ở server.
  const stateToken = jwt.sign({ state, nonce, cv: codeVerifier, uri }, JWT_SECRET, { expiresIn: STATE_TTL_SEC });
  return { url: `${authorizeEndpoint(cfg.tenantId)}?${params}`, stateToken };
}

// ---- Bước 2: đổi `code` lấy id_token, rồi KIỂM id_token ----
async function exchangeAndVerify(req, { code, state }) {
  const cfg = await ssoConfig();
  if (!cfg.enabled) throw Object.assign(new Error('Đăng nhập Microsoft chưa được cấu hình'), { status: 503 });

  const raw = (req.headers.cookie || '').match(/(?:^|;\s*)ktx_sso=([^;]+)/);
  if (!raw) throw Object.assign(new Error('Phiên đăng nhập Microsoft đã hết hạn. Vui lòng thử lại.'), { status: 400 });
  let st;
  try { st = jwt.verify(decodeURIComponent(raw[1]), JWT_SECRET); }
  catch (e) { throw Object.assign(new Error('Phiên đăng nhập Microsoft không hợp lệ hoặc đã hết hạn.'), { status: 400 }); }
  // So state bằng hàm chống dò thời gian (timing-safe)
  const a = Buffer.from(String(state || '')), b = Buffer.from(String(st.state || ''));
  if (a.length !== b.length || !crypto.timingSafeEqual(a, b)) {
    throw Object.assign(new Error('Yêu cầu đăng nhập không khớp (state). Vui lòng thử lại.'), { status: 400 });
  }

  const body = new URLSearchParams({
    client_id: cfg.clientId,
    client_secret: cfg.clientSecret,
    grant_type: 'authorization_code',
    code,
    redirect_uri: st.uri,
    code_verifier: st.cv,
    scope: 'openid profile email',
  });
  const r = await fetch(tokenEndpoint(cfg.tenantId), {
    method: 'POST', headers: { 'Content-Type': 'application/x-www-form-urlencoded' }, body,
  });
  const tok = await r.json().catch(() => ({}));
  if (!r.ok || !tok.id_token) {
    // KHÔNG ném nguyên lỗi của Microsoft ra người dùng (có thể chứa chi tiết cấu hình nội bộ)
    console.error('[SSO] đổi mã thất bại:', r.status, tok.error, tok.error_description);
    throw Object.assign(new Error('Không đổi được mã đăng nhập với Microsoft.'), { status: 502 });
  }

  const decoded = jwt.decode(tok.id_token, { complete: true });
  if (!decoded || !decoded.header || !decoded.header.kid) throw Object.assign(new Error('id_token không hợp lệ'), { status: 400 });
  const key = await getSigningKey(cfg.tenantId, decoded.header.kid);
  let claims;
  try {
    claims = jwt.verify(tok.id_token, key, {
      algorithms: ['RS256'],                    // KHOÁ thuật toán — chặn tấn công đổi alg sang HS256/none
      issuer: issuerOf(cfg.tenantId),
      audience: cfg.clientId,
      clockTolerance: 60,
    });
  } catch (e) {
    throw Object.assign(new Error('Chữ ký id_token không hợp lệ: ' + e.message), { status: 401 });
  }
  if (claims.nonce !== st.nonce) throw Object.assign(new Error('id_token không khớp yêu cầu (nonce).'), { status: 401 });
  // Chặn tài khoản Microsoft NGOÀI tenant công ty (dù iss đã ràng, kiểm thêm tid cho chắc)
  if (claims.tid && claims.tid !== cfg.tenantId) throw Object.assign(new Error('Tài khoản không thuộc tổ chức được phép.'), { status: 403 });

  const email = String(claims.email || claims.preferred_username || '').trim().toLowerCase();
  if (!email) throw Object.assign(new Error('Tài khoản Microsoft không có email — không liên kết được.'), { status: 400 });
  if (cfg.allowedDomains.length) {
    const dom = email.split('@')[1] || '';
    if (!cfg.allowedDomains.includes(dom)) {
      throw Object.assign(new Error(`Email "${email}" không thuộc tên miền được phép đăng nhập.`), { status: 403 });
    }
  }
  return {
    subject: String(claims.oid || claims.sub),  // oid = ổn định theo người trong tenant; email đổi được
    email,
    fullName: String(claims.name || '').trim(),
  };
}

module.exports = { ssoConfig, buildAuthRequest, exchangeAndVerify, redirectUri, STATE_COOKIE, STATE_TTL_SEC };

File: internal/sso/sso.go
// Package sso — đăng nhập Microsoft (OpenID Connect: Authorization Code + PKCE + state/nonce + verify RS256).
// Port từ server/sso.js. state/nonce/code_verifier gói trong JWT ngắn hạn ở cookie httpOnly ktx_sso (stateless).
package sso

import (
	"context"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"ktx/internal/db"
)

const (
	StateCookie = "ktx_sso"
	StateTTLSec = 600 // 10 phút
)

func b64url(b []byte) string { return base64.RawURLEncoding.EncodeToString(b) }
func randBytes(n int) []byte {
	b := make([]byte, n)
	_, _ = rand.Read(b)
	return b
}

// HTTPError: lỗi có status cho handler.
type HTTPError struct {
	Status int
	Msg    string
}

func (e *HTTPError) Error() string { return e.Msg }

type Manager struct {
	secret []byte
	db     *db.DB
}

func NewManager(secret string, database *db.DB) *Manager {
	return &Manager{secret: []byte(secret), db: database}
}

type Config struct {
	Enabled        bool
	TenantID       string
	ClientID       string
	ClientSecret   string
	AllowedDomains []string
}

var envMap2 = map[string]string{
	"sso_tenant_id": "AZURE_TENANT_ID", "sso_client_id": "AZURE_CLIENT_ID",
	"sso_client_secret": "AZURE_CLIENT_SECRET", "sso_allowed_domains": "SSO_ALLOWED_DOMAINS", "sso_enabled": "SSO_ENABLED",
}

// Config: ENV ưu tiên hơn CSDL. server/sso.js:26-38
func (m *Manager) Config(ctx context.Context) Config {
	s, _ := m.db.GetSettings(ctx)
	pick := func(key string) string {
		if env := os.Getenv(envMap2[key]); env != "" {
			return strings.TrimSpace(env)
		}
		return strings.TrimSpace(s[key])
	}
	tenant := pick("sso_tenant_id")
	client := pick("sso_client_id")
	secret := pick("sso_client_secret")
	var domains []string
	for _, d := range strings.Split(pick("sso_allowed_domains"), ",") {
		d = strings.TrimPrefix(strings.ToLower(strings.TrimSpace(d)), "@")
		if d != "" {
			domains = append(domains, d)
		}
	}
	on := pick("sso_enabled") == "true"
	// Secret KHÔNG bắt buộc: có secret -> client tin cậy (confidential); bỏ trống -> public client,
	// đăng nhập dựa PKCE (đã dùng ở BuildAuthRequest/ExchangeAndVerify). Muốn chạy không secret thì
	// app trên Azure phải bật "Allow public client flows".
	return Config{
		Enabled: on && tenant != "" && client != "",
		TenantID: tenant, ClientID: client, ClientSecret: secret, AllowedDomains: domains,
	}
}

// Enabled: cho endpoint /auth/sso/config.
func (m *Manager) Enabled(ctx context.Context) bool { return m.Config(ctx).Enabled }

func issuerOf(t string) string    { return "https://login.microsoftonline.com/" + t + "/v2.0" }
func authorizeEP(t string) string { return "https://login.microsoftonline.com/" + t + "/oauth2/v2.0/authorize" }
func tokenEP(t string) string     { return "https://login.microsoftonline.com/" + t + "/oauth2/v2.0/token" }
func jwksURI(t string) string     { return "https://login.microsoftonline.com/" + t + "/discovery/v2.0/keys" }

// stateClaims: nội dung JWT cookie ktx_sso.
type stateClaims struct {
	State string `json:"state"`
	Nonce string `json:"nonce"`
	CV    string `json:"cv"`
	URI   string `json:"uri"`
	jwt.RegisteredClaims
}

// BuildAuthRequest: dựng URL đẩy sang Microsoft + stateToken (cookie). server/sso.js:74-95
func (m *Manager) BuildAuthRequest(ctx context.Context, redirectURI string) (string, string, error) {
	cfg := m.Config(ctx)
	if !cfg.Enabled {
		return "", "", &HTTPError{503, "Đăng nhập Microsoft chưa được cấu hình"}
	}
	state := b64url(randBytes(24))
	nonce := b64url(randBytes(24))
	cv := b64url(randBytes(48))
	sum := sha256.Sum256([]byte(cv))
	challenge := b64url(sum[:])
	params := url.Values{
		"client_id":             {cfg.ClientID},
		"response_type":         {"code"},
		"redirect_uri":          {redirectURI},
		"response_mode":         {"query"},
		"scope":                 {"openid profile email"},
		"state":                 {state},
		"nonce":                 {nonce},
		"code_challenge":        {challenge},
		"code_challenge_method": {"S256"},
	}
	tok := jwt.NewWithClaims(jwt.SigningMethodHS256, stateClaims{
		State: state, Nonce: nonce, CV: cv, URI: redirectURI,
		RegisteredClaims: jwt.RegisteredClaims{ExpiresAt: jwt.NewNumericDate(time.Now().Add(StateTTLSec * time.Second))},
	})
	stateToken, err := tok.SignedString(m.secret)
	if err != nil {
		return "", "", err
	}
	return authorizeEP(cfg.TenantID) + "?" + params.Encode(), stateToken, nil
}

type Identity struct {
	Subject  string
	Email    string
	FullName string
}

// ---- JWKS cache ----
var (
	jwksMu    sync.Mutex
	jwksTID   string
	jwksAt    time.Time
	jwksKeys  []jwkKey
	httpCl    = &http.Client{Timeout: 10 * time.Second}
)

type jwkKey struct {
	Kid string `json:"kid"`
	N   string `json:"n"`
	E   string `json:"e"`
	Kty string `json:"kty"`
}

func fetchJWKS(ctx context.Context, tenant string) ([]jwkKey, error) {
	req, _ := http.NewRequestWithContext(ctx, "GET", jwksURI(tenant), nil)
	r, err := httpCl.Do(req)
	if err != nil {
		return nil, err
	}
	defer r.Body.Close()
	if r.StatusCode != 200 {
		return nil, fmt.Errorf("Không lấy được khoá công khai của Microsoft (JWKS)")
	}
	var j struct {
		Keys []jwkKey `json:"keys"`
	}
	if err := json.NewDecoder(r.Body).Decode(&j); err != nil {
		return nil, err
	}
	return j.Keys, nil
}

func (m *Manager) signingKey(ctx context.Context, tenant, kid string) (*rsa.PublicKey, error) {
	jwksMu.Lock()
	fresh := jwksTID == tenant && time.Since(jwksAt) < time.Hour
	keys := jwksKeys
	jwksMu.Unlock()
	find := func(ks []jwkKey) *jwkKey {
		for i := range ks {
			if ks[i].Kid == kid {
				return &ks[i]
			}
		}
		return nil
	}
	var jwk *jwkKey
	if fresh {
		jwk = find(keys)
	}
	if jwk == nil { // hết hạn hoặc xoay khoá -> nạp lại
		ks, err := fetchJWKS(ctx, tenant)
		if err != nil {
			return nil, err
		}
		jwksMu.Lock()
		jwksTID, jwksAt, jwksKeys = tenant, time.Now(), ks
		jwksMu.Unlock()
		jwk = find(ks)
	}
	if jwk == nil {
		return nil, fmt.Errorf("Không tìm thấy khoá ký khớp id_token")
	}
	return jwkToRSA(jwk)
}

func jwkToRSA(k *jwkKey) (*rsa.PublicKey, error) {
	nb, err := base64.RawURLEncoding.DecodeString(k.N)
	if err != nil {
		return nil, err
	}
	eb, err := base64.RawURLEncoding.DecodeString(k.E)
	if err != nil {
		return nil, err
	}
	e := 0
	for _, b := range eb {
		e = e<<8 | int(b)
	}
	return &rsa.PublicKey{N: new(big.Int).SetBytes(nb), E: e}, nil
}

// ExchangeAndVerify: đổi code lấy id_token rồi KIỂM. server/sso.js:98-163
func (m *Manager) ExchangeAndVerify(ctx context.Context, ssoCookie, code, state string) (Identity, error) {
	cfg := m.Config(ctx)
	if !cfg.Enabled {
		return Identity{}, &HTTPError{503, "Đăng nhập Microsoft chưa được cấu hình"}
	}
	if ssoCookie == "" {
		return Identity{}, &HTTPError{400, "Phiên đăng nhập Microsoft đã hết hạn. Vui lòng thử lại."}
	}
	var st stateClaims
	_, err := jwt.ParseWithClaims(ssoCookie, &st, func(t *jwt.Token) (interface{}, error) { return m.secret, nil }, jwt.WithValidMethods([]string{"HS256"}))
	if err != nil {
		return Identity{}, &HTTPError{400, "Phiên đăng nhập Microsoft không hợp lệ hoặc đã hết hạn."}
	}
	if subtle.ConstantTimeCompare([]byte(state), []byte(st.State)) != 1 {
		return Identity{}, &HTTPError{400, "Yêu cầu đăng nhập không khớp (state). Vui lòng thử lại."}
	}

	form := url.Values{
		"client_id":  {cfg.ClientID},
		"grant_type": {"authorization_code"}, "code": {code},
		"redirect_uri": {st.URI}, "code_verifier": {st.CV}, "scope": {"openid profile email"},
	}
	// Có secret -> gửi kèm (confidential). Bỏ trống -> KHÔNG gửi, chỉ dựa PKCE (public client).
	if cfg.ClientSecret != "" {
		form.Set("client_secret", cfg.ClientSecret)
	}
	req, _ := http.NewRequestWithContext(ctx, "POST", tokenEP(cfg.TenantID), strings.NewReader(form.Encode()))
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	r, err := httpCl.Do(req)
	if err != nil {
		return Identity{}, &HTTPError{502, "Không đổi được mã đăng nhập với Microsoft."}
	}
	defer r.Body.Close()
	body, _ := io.ReadAll(r.Body)
	var tok struct {
		IDToken string `json:"id_token"`
	}
	_ = json.Unmarshal(body, &tok)
	if r.StatusCode != 200 || tok.IDToken == "" {
		// NÊU RÕ lý do Microsoft từ chối (vd AADSTS7000218 = app kiểu Web bắt buộc client_secret) để
		// admin biết đường sửa, thay vì thông báo chung chung.
		var e struct {
			Error string `json:"error"`
			Desc  string `json:"error_description"`
		}
		_ = json.Unmarshal(body, &e)
		reason := e.Desc
		if reason == "" {
			reason = e.Error
		}
		if i := strings.IndexAny(reason, "\r\n"); i > 0 { // Microsoft trả nhiều dòng -> lấy dòng đầu
			reason = reason[:i]
		}
		if len(reason) > 220 {
			reason = reason[:220]
		}
		msg := "Microsoft từ chối đổi mã đăng nhập."
		if reason != "" {
			msg += " Lý do: " + reason
		}
		return Identity{}, &HTTPError{502, msg}
	}

	return m.VerifyIDToken(ctx, tok.IDToken, st.Nonce)
}

// VerifyIDToken: KIỂM id_token (chữ ký JWKS RS256 + iss + aud=client_id + tid + tên miền cho phép).
// Dùng chung cho: (1) luồng server-side ExchangeAndVerify (truyền nonce từ state cookie), và (2) luồng
// SPA — trình duyệt tự đổi mã rồi gửi id_token về /sso/verify (nonce đã kiểm phía trình duyệt -> "").
// KHÔNG đổi mã, KHÔNG cần client_secret: chỉ tin token do Microsoft ký cho ĐÚNG app + ĐÚNG tenant.
func (m *Manager) VerifyIDToken(ctx context.Context, idToken, expectedNonce string) (Identity, error) {
	cfg := m.Config(ctx)
	if !cfg.Enabled {
		return Identity{}, &HTTPError{503, "Đăng nhập Microsoft chưa được cấu hình"}
	}
	var claims jwt.MapClaims
	_, err := jwt.ParseWithClaims(idToken, &claims, func(t *jwt.Token) (interface{}, error) {
		kid, _ := t.Header["kid"].(string)
		if kid == "" {
			return nil, fmt.Errorf("id_token không hợp lệ")
		}
		return m.signingKey(ctx, cfg.TenantID, kid)
	}, jwt.WithValidMethods([]string{"RS256"}), jwt.WithIssuer(issuerOf(cfg.TenantID)),
		jwt.WithAudience(cfg.ClientID), jwt.WithLeeway(60*time.Second))
	if err != nil {
		return Identity{}, &HTTPError{401, "Chữ ký id_token không hợp lệ: " + err.Error()}
	}
	if expectedNonce != "" {
		if nonce, _ := claims["nonce"].(string); nonce != expectedNonce {
			return Identity{}, &HTTPError{401, "id_token không khớp yêu cầu (nonce)."}
		}
	}
	if tid, _ := claims["tid"].(string); tid != "" && tid != cfg.TenantID {
		return Identity{}, &HTTPError{403, "Tài khoản không thuộc tổ chức được phép."}
	}
	email := strings.ToLower(strings.TrimSpace(claimStr(claims, "email", "preferred_username")))
	if email == "" {
		return Identity{}, &HTTPError{400, "Tài khoản Microsoft không có email — không liên kết được."}
	}
	if len(cfg.AllowedDomains) > 0 {
		dom := ""
		if i := strings.LastIndex(email, "@"); i >= 0 {
			dom = email[i+1:]
		}
		ok := false
		for _, d := range cfg.AllowedDomains {
			if d == dom {
				ok = true
				break
			}
		}
		if !ok {
			return Identity{}, &HTTPError{403, `Email "` + email + `" không thuộc tên miền được phép đăng nhập.`}
		}
	}
	subject := claimStr(claims, "oid", "sub")
	return Identity{Subject: subject, Email: email, FullName: strings.TrimSpace(claimStr(claims, "name"))}, nil
}

func claimStr(c jwt.MapClaims, keys ...string) string {
	for _, k := range keys {
		if v, ok := c[k].(string); ok && v != "" {
			return v
		}
	}
	return ""
}
